### PR TITLE
Allow override of checkout process

### LINKED ISSUE #19848: Allow override of checkoutProcess construction
#### Is your feature request related to a problem?

I would like to modify the content of `controllers/front/OrderController.php` property `$checkoutProcess` when it is is constructed in `bootstrap()` function but I cannot do it through an override, it's not available.

#### Describe the solution you'd like

Put the logic of construction of `$checkoutProcess` inside a protected function to allow developers to override it

#### Alternatives you've considered

Use a hook

### PR DESCRIPTION
**Pull request by ksaandev**
To modify $this->checkoutProcess I need to be able to override it

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Move code to create `$this->checkoutProcess` into protected function to override it
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/19848
| How to test?  | Use provided test module (see below). Install it and open Checkout page, you should see 2 "delivery steps" instead of one !

Test module - v2
[testmodule.zip](https://github.com/PrestaShop/PrestaShop/files/4931761/testmodule.zip)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/19830)

### CODE REVIEW

**Comment by NeOMakinG** [comment]
Hey @ksaandev, looks like there is a problem with CSFixer, also, would it be possible that you create an issue for this ? It's necessary for the changelog and some things :/

Also, I think the QA needs to test it, maybe we need a module for them !

wdyt @PrestaShop/prestashop-core-developers

**Comment by ksaandev** [comment]
> Hey @ksaandev, looks like there is a problem with CSFixer, also, would it be possible that you create an issue for this ? It's necessary for the changelog and some things :/
> 
> Also, I think the QA needs to test it, maybe we need a module for them !
> 
> wdyt @PrestaShop/prestashop-core-developers

Issue created https://github.com/PrestaShop/PrestaShop/issues/19848

Fixed PHP CS Fixer

**Comment by ksaandev** [comment]
I cannot give you my module to test because it is paying.

Maybe I can make small module for demo

**Comment by matks** [comment]
Hi @ksaandev, PHP-CS-Fixer is not happy can you fix it ?

**Comment by ksaandev** [comment]
> Hi @ksaandev, PHP-CS-Fixer is not happy can you fix it ?

Done !

File: controllers/front/OrderController.php [hunk 1]
```diff
@@ -113,60 +113,11 @@ public function getCheckoutSession()
...
-                $this->context,
-                $translator,
-                new PaymentOptionsFinder(),
-                new ConditionsToApproveFinder(
-                    $this->context,
-                    $translator
-                )
-            ));
+        $this->checkoutProcess = $this->buildCheckoutProcess($session, $translator);
+
```
Comment (matks) [inline controllers/front/OrderController.php]:
```suggestion

```

**Comment by matks** [comment]
> > Hi @ksaandev, PHP-CS-Fixer is not happy can you fix it ?
> 
> Done !

Hi, I'm sorry 😅 php-cs-fixer is still red, there's an extra blank character. I will remove it 😉

**Comment by matks** [comment]
@ksaandev PHP-CS-Fixer is happy now 😄 

I think this PR is valid, however we need
- to put this new hook Interesting into file install-dev/data/xml/hook.xml ; this is a registry for all new hooks
- insert your new hook into dedicated SQL table through the file install-dev/upgrade/sql/1.7.8.0.sql
- provide a small demo module to help QA test your PR

Please tell us if something is not clear in these needs 😉

**Comment by ksaandev** [comment]
> @ksaandev PHP-CS-Fixer is happy now 😄
> 
> I think this PR is valid, however we need
> 
> * to put this new hook Interesting into file install-dev/data/xml/hook.xml ; this is a registry for all new hooks
> * insert your new hook into dedicated SQL table through the file install-dev/upgrade/sql/1.7.8.0.sql
> * provide a small demo module to help QA test your PR
> 
> Please tell us if something is not clear in these needs 😉

Hi ! I did all that and I added a test module in my 1st message in this PR

I also added a new function in CheckoutProcess to override steps which allows to insert new steps in middle of process (not only at the end), see how it works:
```
public function hookActionCheckoutRender($params)
    {
        $this->translator = Context::getContext()->getTranslator();
        /** @var CheckoutProcess $process */
        $process = $params['checkoutProcess'];
        $steps = $process->getSteps();

        $checkoutDeliveryStep = new CheckoutDeliveryStep(
            $this->context,
            $this->translator
        );
        $checkoutDeliveryStep->setCheckoutProcess($process);

        $newSteps = [
            $steps[0],
            $steps[1],
            $checkoutDeliveryStep,
            $steps[2],
            $steps[3]
        ];

        $process->overrideSteps($newSteps);
    }
```

**Comment by ksaandev** [comment]
Hi, I have also fix some git issues conflict

**Review by matks — DISMISSED** [review]
This seems good, thank you @ksaandev. I approve this PR 👍

File: classes/checkout/CheckoutProcess.php [hunk 2]
```diff
@@ -108,6 +108,14 @@ public function getSteps()
         return $this->steps;
     }
 
+    /**
+     * @param array $steps
+     */
+    public function overrideSteps(array $steps)
```
Comment (PierreRambaud) [inline classes/checkout/CheckoutProcess.php]:
```suggestion
    public function setSteps(array $steps)
```

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
As we use `addStep`, `getSteps` you must rename your method `setSteps` to be consistent.

**Comment by PierreRambaud** [comment]
Only one thing and it's ok for me :)

**Comment by ksaandev** [comment]
@PierreRambaud thank you ! I applied your suggestion

**Review by PierreRambaud — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by SD1982** [comment]
Hi @ksaandev I have an exception during the checkout process and it seems to be related to the provided module can you check this please ?
https://drive.google.com/file/d/1PWpffnZhQUkq9BlYn32Bkmwy0o48boZ8/view?usp=sharing

**Comment by ksaandev** [comment]
> Hi @ksaandev I have an exception during the checkout process and it seems to be related to the provided module can you check this please ?
> https://drive.google.com/file/d/1PWpffnZhQUkq9BlYn32Bkmwy0o48boZ8/view?usp=sharing

Hi ! Yes you are right. Sorry about that, it's because we changed the code of the PR here https://github.com/PrestaShop/PrestaShop/pull/19830/commits/ddd0e697b37d1d6bef6e2842047102833406ec2f but not in my module. I will send you a new version of the module today

**Comment by SD1982** [comment]
Hi @ksaandev
Don't hesitate to ping me when you have updated the module and I will treat your PR as soon as it has been reviewed ! :+1:

**Comment by ksaandev** [comment]
> Hi @ksaandev
> Don't hesitate to ping me when you have updated the module and I will treat your PR as soon as it has been reviewed ! 👍

I have updated my module to "Test module - v2" (see the PR description) to be synchronized with latest changes. Sorry for the error. I have checked on my shop and it works fine. Dont hesitate to call me if something bad

Thanks

**Review by matks — APPROVED** [review]

**Comment by SD1982** [comment]
Hi @ksaandev it 's me again ! 
The exception seems to not appear anymore but the checkout process is locked on the third step 
When I click on the continue button the page is refreshed and nothing else happen 
https://drive.google.com/file/d/1Fvs9NhlmItzTp7KHo5buKWYdVtUz0bzc/view?usp=sharing

**Comment by ksaandev** [comment]
> Hi @ksaandev it 's me again !
> The exception seems to not appear anymore but the checkout process is locked on the third step
> When I click on the continue button the page is refreshed and nothing else happen
> https://drive.google.com/file/d/1Fvs9NhlmItzTp7KHo5buKWYdVtUz0bzc/view?usp=sharing

Yes, it's because it's just to show it appears. It would need a big module to add a real step inside the process. The PR only adds the possibility for hook to add it, it is responsibility of module to then inject something good.

**Comment by SD1982** [comment]
Hello @ksaandev  Would it be possible to have a module that would allow us to check that the hook works correctly while allowing to go to the end of the command process?

**Comment by PierreRambaud** [comment]
@SD1982 you can use https://github.com/PrestaShop/ps_qualityassurance and use the `actionCheckoutRender` hook :wink:

**Comment by ksaandev** [comment]
Hi I am not sure https://github.com/PrestaShop/ps_qualityassurance can provide what @SD1982 wants.

Please try https://github.com/PrestaShop/ps_qualityassurance and tell me if it is good. If it is no good I will try make module for real checkout step but it is very long.

This PR only adds the step, it does not modify how checkout works. I am not sure module with complete step is necessary

**Comment by SD1982** [comment]
Hi @ksaandev  I have tested with https://github.com/PrestaShop/ps_qualityassurance  but i have a problem similar to your module i can"t go to the end of the checkout process . When the hook is triggered the page become blank and i can't continue the checkout 
https://drive.google.com/file/d/1og1YpGucYkDwlgNkQ6Wm3Dlu5LYSbbV5/view?usp=sharing

**Comment by ksaandev** [comment]
> Hi @ksaandev I have tested with https://github.com/PrestaShop/ps_qualityassurance but i have a problem similar to your module i can"t go to the end of the checkout process . When the hook is triggered the page become blank and i can't continue the checkout
> https://drive.google.com/file/d/1og1YpGucYkDwlgNkQ6Wm3Dlu5LYSbbV5/view?usp=sharing

I know I said it 😄  https://github.com/PrestaShop/PrestaShop/pull/19830#issuecomment-662110890 

OK I will do it but it is going to take long time. I dont have lot of free time now maybe in some weeks.

**Comment by SD1982** [comment]
Hi @ksaandev 
Good news !! After some few additional tests and the help of @PierreRambaud and the new [QA module](https://github.com/PrestaShop/ps_qualityassurance) I was able to correctly test your PR and correctly trigger the hook and go without error until the end of the command process .
So you don"t need to work to provide a new module !
Thanks @ksaandev !!

**Comment by Progi1984** [comment]
Thanks @ksaandev & @SD1982

**Comment by ksaandev** [comment]
Thank you guys 😁

**Pull request merged**